server/server.go

```diff
@@ -55,18 +55,34 @@ func StartServer(cfg *config.Config) error {
 	}
 
 	if cfg.CoreServiceMode {
-		rpcServer := rpccore.NewRPCServer(sig, cfg.DataDir)
+		rpcServer := rpccore.NewRPCServer(stop, cfg.DataDir)
 		grpcServer = grpc.NewServer(options...)
 
 		apicore.RegisterPoetCoreProverServer(grpcServer, rpcServer)
 		apicore.RegisterPoetVerifierServer(grpcServer, rpcServer)
 		proxyRegstr = append(proxyRegstr, apicore.RegisterPoetCoreProverHandlerFromEndpoint)
 		proxyRegstr = append(proxyRegstr, apicore.RegisterPoetVerifierHandlerFromEndpoint)
 	} else {
-		svc, err := service.NewService(sig, cfg.Service, cfg.DataDir)
+		svc, err := service.NewService(serverGroup, cfg.Service, cfg.DataDir)
```

Comment:
imho `NewService` should not receive the errGroup as parameter, but rather just the context and use it to create its own `errGroup`. goroutines started by different components share the same `errGroup`, which could lead to problems by components interfering with each other.

Reply:
The reason to share this error group is to provide a graceful shutdown, which would allow all goroutines to finish. It's required to let a certain state to be persisted when the application closes (when the context is cancelled). 

Previously it was achieved via calls to `BlockShutdown`. Now, this synchronization is done via calling `.Wait()` on the errgroup.

Reply:
I am aware 🙂, but this way if a routine by `service` errors the `server` also shuts down. Previously this would only happen on a shutdown signal.

Instead I think errors should bubble up and the parent should decide if they want to restart the service or also shutdown.

Reply:
Refactored. `Service` now holds its own `errgroup` inside, similar to `http.Server` from the stdlib.